### ISSUE TITLE
debootstrap, permissions issue -- difficult to implicate nodev mount option of /tmp/

### ISSUE THREAD

**Issue by yarikoptic** [apptainer/singularity#1285]
## Version of Singularity:

```shell
$> apt-cache policy singularity-container
singularity-container:
  Installed: 2.4.1-1~nd90+1
  Candidate: 2.4.1-1~nd90+1
  Version table:
 *** 2.4.1-1~nd90+1 500
        500 http://neuro.debian.net/debian stretch/main amd64 Packages
        500 http://neurodebian.ovgu.de/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
```

### Expected behavior

build / debootstrap doesn't fail

### Actual behavior

fails while complaining

```shell
/usr/sbin/debootstrap: 1454: /usr/sbin/debootstrap: cannot create /var/lib/singularity/mnt/container/test-dev-null: Permission denied
E: Cannot install into target '/var/lib/singularity/mnt/container' mounted with noexec or nodev
```

### Steps to reproduce behavior

I have a minimalistic def file:
```shell
$> cat sing.def 
BootStrap: debootstrap
OSVersion: stretch
MirrorURL: http://ftp.us.debian.org/debian/
```
and invoking build under /tmp where that file is leads to:

```shell
$> sudo singularity --debug build test.img sing.def 2>&1 | tee test.log
[sudo] password for yoh: 
Enabling debugging
Ending argument loop
Singularity version: 2.4.1-dist
Exec'ing: /usr/lib/x86_64-linux-gnu/singularity/cli/build.exec
Using container recipe deffile: sing.def
VERBOSE [U=0,P=3019271]    message_init()                            Set messagelevel to: 5
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Initialize configuration file: /etc/singularity/singularity.conf
DEBUG   [U=0,P=3019271]    singularity_config_parse()                Starting parse of configuration file /etc/singularity/singularity.conf
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key allow setuid = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key max loop devices = '256'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key allow pid ns = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key config passwd = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key config group = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key config resolv_conf = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount proc = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount sys = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount dev = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount devpts = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount home = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount tmp = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount hostfs = 'no'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key bind path = '/etc/localtime'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key bind path = '/etc/hosts'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key user bind control = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key enable overlay = 'try'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key mount slave = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key sessiondir max size = '16'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key allow container squashfs = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key allow container extfs = 'yes'
VERBOSE [U=0,P=3019271]    singularity_config_parse()                Got config key allow container dir = 'yes'
DEBUG   [U=0,P=3019271]    singularity_config_parse()                Finished parsing configuration file '/etc/singularity/singularity.conf'
VERBOSE [U=0,P=3019271]    singularity_registry_init()               Initializing Singularity Registry
VERBOSE [U=0,P=3019271]    singularity_registry_set()                Adding value to registry: 'COMMAND' = 'build'
DEBUG   [U=0,P=3019271]    singularity_registry_set()                Returning singularity_registry_set(COMMAND, build) = 0
VERBOSE [U=0,P=3019271]    singularity_registry_set()                Adding value to registry: 'SYSCONFDIR' = '/etc'
DEBUG   [U=0,P=3019271]    singularity_registry_set()                Returning singularity_registry_set(sysconfdir, /etc) = 0
VERBOSE [U=0,P=3019271]    singularity_registry_set()                Adding value to registry: 'BUILDDEF' = 'sing.def'
DEBUG   [U=0,P=3019271]    singularity_registry_set()                Returning singularity_registry_set(BUILDDEF, sing.def) = 0
VERBOSE [U=0,P=3019271]    singularity_registry_set()                Adding value to registry: 'ROOTFS' = '/tmp/.singularity-build.leQw88'
DEBUG   [U=0,P=3019271]    singularity_registry_set()                Returning singularity_registry_set(ROOTFS, /tmp/.singularity-build.leQw88) = 0
VERBOSE [U=0,P=3019271]    singularity_registry_set()                Adding value to registry: 'VERSION' = '2.4.1-dist'
DEBUG   [U=0,P=3019271]    singularity_registry_set()                Returning singularity_registry_set(version, 2.4.1-dist) = 0
...
VERBOSE [U=0,P=3019283]    message_init()                            Set messagelevel to: 5
DEBUG   [U=0,P=3019283]    main()                                    Iterating through file looking for sections matching: %pre
Adding base Singularity environment to container
tar: ./.exec: implausibly old time stamp -9223372036854775808
tar: ./.run: implausibly old time stamp -9223372036854775808
tar: ./.shell: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/exec: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/run: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/shell: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/start: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/test: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/01-base.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/90-environment.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/95-apps.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/99-base.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/libs: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/runscript: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/startscript: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d: implausibly old time stamp -9223372036854775808
tar: ./.test: implausibly old time stamp -9223372036854775808
tar: ./dev: implausibly old time stamp -9223372036854775808
tar: ./environment: implausibly old time stamp -9223372036854775808
tar: ./etc/hosts: implausibly old time stamp -9223372036854775808
tar: ./etc/resolv.conf: implausibly old time stamp -9223372036854775808
tar: ./etc: implausibly old time stamp -9223372036854775808
tar: ./home: implausibly old time stamp -9223372036854775808
tar: ./proc: implausibly old time stamp -9223372036854775808
tar: ./root: implausibly old time stamp -9223372036854775808
tar: ./singularity: implausibly old time stamp -9223372036854775808
tar: ./sys: implausibly old time stamp -9223372036854775808
tar: ./tmp: implausibly old time stamp -9223372036854775808
tar: ./var/tmp: implausibly old time stamp -9223372036854775808
tar: ./var: implausibly old time stamp -9223372036854775808
tar: .: implausibly old time stamp -9223372036854775808
/usr/sbin/debootstrap: 1454: /usr/sbin/debootstrap: cannot create /var/lib/singularity/mnt/container/test-dev-null: Permission denied
E: Cannot install into target '/var/lib/singularity/mnt/container' mounted with noexec or nodev
ABORT: Aborting with RETVAL=255
Cleaning up...
```

unfortunately `--debug` output does not include debootstrap call details:
```shell

$> grep debootstrap test.log
VERBOSE [U=0,P=3019271]    main()                                    Got bootstrap definition key/val 'BootStrap' = 'debootstrap'
VERBOSE [U=0,P=3019271]    singularity_registry_set()                Adding value to registry: 'DRIVER' = 'debootstrap'
DEBUG   [U=0,P=3019271]    singularity_registry_set()                Returning singularity_registry_set(DRIVER, debootstrap) = 0
DEBUG   [U=0,P=3019271]    envar_set()                               Setting environment variable: 'BOOTSTRAP' = 'debootstrap'
DEBUG   [U=0,P=3019271]    envar_set()                               Setting environment variable: 'SINGULARITY_DEFFILE_BOOTSTRAP' = 'debootstrap'
/usr/sbin/debootstrap: 1454: /usr/sbin/debootstrap: cannot create /var/lib/singularity/mnt/container/test-dev-null: Permission denied
```
Note that the error is about `/var/lib/singularity/mnt/container/` location which had no special mount options relevant.
Only after seeing above references to /tmp/ being used, I have checked for its mount options to include `nodev`, and remounting it without `nodev` resolved it.
Also -- I believe that I have not experienced this issue with previous singularity 2.2.

It would be nice if
- [ ] `--debug` output provided details of call to bootstrap (directory under which it is executed and all cmdline options)
- [ ] avoided requiring a temporary directory to support devices (security implications)

**Comment by verdurin** [comment]
Just saw this on Fedora 27 with Singularity 2.4.5.

**Comment by doublesharp** [comment]
@yarikoptic Is your home directory encrypted? I was able to solve this same error by remounting my user directory with a `-i`.

**Error**
```
USERNAME@host:~/gitian-builder$ bin/make-base-vm --suite trusty --arch amd64 --lxc
/usr/sbin/debootstrap: 1368: /usr/sbin/debootstrap: cannot create /home/USERNAME/builder/base-trusty-amd64-bootstrap/test-dev-null: Permission denied
E: Cannot install into target '/home/USERNAME/builder/base-trusty-amd64-bootstrap' mounted with noexec or nodev
```
**Remount**
```
$ sudo mount -i -o remount,exec,dev /home/USERNAME
```
**Working**
```
$ bin/make-base-vm --suite trusty --arch amd64 --lxc
I: Retrieving InRelease 
I: Failed to retrieve InRelease
I: Retrieving Release 
I: Retrieving Release.gpg 
# ... and so on
```

**Comment by yarikoptic** [comment]
ha @doublesharp -- indeed it is!  But I am not sure why it is of concern since my original message was about /var/lib/singularity/mnt/container and that one is not mounted with "nodev":
```shell
$> df -h /var/lib/singularity/mnt/container
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda1        89G   83G  1.3G  99% /

$> mount | grep sda1                       
/dev/sda1 on / type ext4 (rw,relatime,errors=remount-ro,stripe=32719,data=ordered)
/dev/sda1 on /var/lib/docker/devicemapper type ext4 (rw,relatime,errors=remount-ro,stripe=32719,data=ordered)
```
Also I guess remounting with -i  might have some undesirable effects since it is just for
```
       -i, --internal-only
              Don't call the /sbin/mount.filesystem helper even if it exists.
```

**Comment by dtrudg** [comment]
Closing this as there hasn't been any update on newer versions of Singularity. If the issue exsits in a current 3.x it'd be great to have a new issue opened.
